server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java

```diff
@@ -359,169 +352,117 @@ protected ClusterStateUpdateResponse newResponse(boolean acknowledged) {
+            if (metadata.hasIndex(resizedIndexName)) {
+                throw new ResourceAlreadyExistsException(resizedIndexName);
+            }
+            if (!currentState.blocks().indexBlocked(ClusterBlockLevel.WRITE, sourceIndexName)) {
+                throw new IllegalStateException("index " + sourceIndex + " must be read-only to resize index. use \"index.blocks.write=true\"");
             }
-            assert template == null : String.format(Locale.ENGLISH, "Found a matching template for index %s, invalid usage.", request.index());
 
-            final Index recoverFromIndex = request.recoverFrom();
-            MappingMetadata mapping;
-            if (recoverFromIndex == null) {
-                mapping = new MappingMetadata(Map.of());
+            final int routingNumShards;
+            Settings sourceSettings = sourceIndex.getSettings();
+            Version indexVersionCreated = currentState.nodes().getSmallestNonClientNodeVersion();
+            int newNumShards = request.newNumShards();
+            if (sourceIndex.getNumberOfShards() == 1) {
+                routingNumShards = IndexMetadata.INDEX_NUMBER_OF_ROUTING_SHARDS_SETTING.exists(sourceSettings)
+                    ? IndexMetadata.INDEX_NUMBER_OF_ROUTING_SHARDS_SETTING.get(sourceSettings)
+                    : calculateNumRoutingShards(newNumShards, indexVersionCreated);
             } else {
-                IndexMetadata sourceMetadata = currentState.metadata().getIndexSafe(recoverFromIndex);
-                mapping = sourceMetadata.mapping();
+                routingNumShards = sourceIndex.getRoutingNumShards();
             }
 
-            Settings.Builder indexSettingsBuilder = Settings.builder()
-                .put(request.settings())
+            Settings.Builder indexSettingsBuilder = Settings.builder();
+            for (final String key : sourceSettings.keySet()) {
+                final Setting<?> setting = indexScopedSettings.get(key);
+                if (setting == null) {
+                    assert indexScopedSettings.isPrivateSetting(key) : key;
+                } else if (setting.getProperties().contains(Setting.Property.NotCopyableOnResize)) {
+                    continue;
+                }
+                indexSettingsBuilder.copy(key, sourceSettings);
+            }
+            indexSettingsBuilder
                 .put(SETTING_INDEX_UUID, UUIDs.randomBase64UUID())
                 .put(SETTING_CREATION_DATE, Instant.now().toEpochMilli())
-                .put(
-                    IndexMetadata.SETTING_INDEX_VERSION_CREATED.getKey(),
-                    currentState.nodes().getSmallestNonClientNodeVersion());
+                .put(IndexMetadata.SETTING_INDEX_VERSION_CREATED.getKey(), indexVersionCreated)
+                .put(IndexMetadata.INDEX_RESIZE_SOURCE_NAME.getKey(), sourceIndexName)
+                .put(IndexMetadata.INDEX_RESIZE_SOURCE_UUID.getKey(), sourceIndex.getIndexUUID())
```

Comment:
Shouldn't these only be set for a *skrink*  operation? At least it looks like it was like this before.

Reply:
Before it set INDEX_SHRINK and INDEX_RESIZE settings. The `INDEX_SHRINK` is more or less deprecated:

https://github.com/crate/crate/blob/d85eba28a369253079aca3676f102996e75fbc96/server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java#L457-L460

If I understand it right we can remove the setting entirely as it was only added in addition so that older nodes that get a resized-shard allocated can recover it correctly. But with 6.0.0 we no longer have any nodes not setting the new `INDEX_RESIZE` settings.

Reply:
My fault, I mixed it up with the deprecated `SHRINK` settings. sorry.

> If I understand it right we can remove the setting entirely as it was only added in addition so that older nodes that get a resized-shard allocated can recover it correctly. But with 6.0.0 we no longer have any nodes not setting the new INDEX_RESIZE settings.
> 
> I can follow up on this in a separate PR

Sounds good, I agree, they can be removed completely.